libkbfs/folder_branch_ops.go

```diff
@@ -6852,12 +6861,178 @@ func (fbo *folderBranchOps) KickoffAllOutstandingRekeys() error {
+	nameToNextPage = make(map[string][]byte, len(convIDs))
+	for i, id := range convIDs {
+		fbo.config.Chat().RegisterForMessages(id, fbo.receiveNewEditChat)
+		name := channelNames[i]
+		idToName[id.String()] = name
+		nameToID[name] = id
+		nextPage := fbo.getEditMessages(ctx, id, name, nil)
+		if nextPage != nil {
+			nameToNextPage[name] = nextPage
+		}
+	}
+	return idToName, nameToID, nameToNextPage
+}
+
+func (fbo *folderBranchOps) getEditMessages(
+	ctx context.Context, id chat1.ConversationID, channelName string,
+	startPage []byte) (nextPage []byte) {
+	// TODO: be smarter about not fetching messages we've already
+	// seen?  `AddNotifications` below will filter out any duplicates,
+	// so it's not strictly needed for correctness.
+	messages, nextPage, err := fbo.config.Chat().ReadChannel(ctx, id, startPage)
+	if err != nil {
+		fbo.log.CWarningf(ctx, "Couldn't get messages for conv %s: %+v",
+			id, err)
+		return nil
+	}
+	err = fbo.editHistory.AddNotifications(channelName, messages)
+	if err != nil {
+		fbo.log.CWarningf(ctx, "Couldn't add messages for conv %s: %+v",
+			id, err)
+		return nil
+	}
+	return nextPage
+}
+
+func (fbo *folderBranchOps) monitorEditsChat() {
+	ctx, cancelFunc := fbo.newCtxWithFBOID()
+	defer cancelFunc()
+	fbo.log.CDebugf(ctx, "Starting kbfs-edits chat monitoring")
+
+	fbo.cancelEditsLock.Lock()
+	fbo.cancelEdits = cancelFunc
+	fbo.cancelEditsLock.Unlock()
+
+	// Register for all the channels of this chat.
+	lState := makeFBOLockState()
+	md, _ := fbo.getHead(lState)
+	name := md.GetTlfHandle().GetCanonicalName()
+
+	idToName, nameToID, nameToNextPage := fbo.initEditChatChannels(ctx, name)
+
+	finishActivity := false
+	for {
+		// Recompute the history, and fetch more messages for any
+		// writers who need them.
+		writersWhoNeedMore := fbo.editHistory.Recompute()
+		gotMore := false
+		for w, needsMore := range writersWhoNeedMore {
+			if !needsMore {
+				continue
+			}
+			if startPage, ok := nameToNextPage[w]; ok && startPage != nil {
+				id, ok := nameToID[w]
+				if !ok {
+					fbo.log.CDebugf(ctx, "No channel found for %s", w)
+					continue
+				}
+				fbo.log.CDebugf(
+					ctx, "Going to fetch more messages for writer %s", w)
+				gotMore = true
+				nextPage := fbo.getEditMessages(ctx, id, w, startPage)
+				if nextPage == nil {
+					delete(nameToNextPage, w)
+				} else {
+					nameToNextPage[w] = nextPage
+				}
+			}
+		}
+		if gotMore {
+			// Retry the recompute, now with more messages.
+			continue
+		}
+
+		// Update the overall user history.  TODO: if the TLF name
+		// changed, we should clean up the old user history.
+		fbo.config.UserHistory().UpdateHistory(
+			name, fbo.id().Type(), fbo.editHistory)
+
+		if finishActivity {
+			fbo.editActivity.Done()
+		}
+
+		select {
+		case <-fbo.shutdownChan:
+			fbo.log.CDebugf(ctx, "Shutting down chat monitoring")
+			return
+		case a := <-fbo.editChannels:
+			finishActivity = true
```

Comment:
We set this to true once and then never reset it to `false`.

IMO it'd be more readable to have this inside a `func()` with `defer fbo.editActivity.Done()`

Reply:
Refactored this whole function to make it cleaner.